common/conf/osscrypt.go

```diff
@@ -0,0 +1,230 @@
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode/100 != 2 {
+		return nil, "", fmt.Errorf("unexpected status code %d from auth", resp.StatusCode)
+	}
+	secretStringBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, "", fmt.Errorf("erroring reading response from auth %s", err)
+	}
+	secret, err := base64.StdEncoding.DecodeString(string(secretStringBytes))
+	if err != nil {
+		return nil, "", fmt.Errorf("stored secret %q set in Barbican is invalid: %s", string(secretStringBytes), err)
+	}
+	return secret, "Barbican", nil
+}
+
+type keystoneRequestV3 struct {
+	Auth struct {
+		Identity struct {
+			Methods  []string `json:"methods"`
+			Password struct {
+				User struct {
+					Name   string `json:"name"`
+					Domain struct {
+						Name string `json:"name"`
+					} `json:"domain"`
+					Password string `json:"password"`
+				} `json:"user"`
+			} `json:"password"`
+		} `json:"identity"`
+	} `json:"auth"`
+}
+
+type keystoneResponseV3 struct {
+	Token struct {
+		Catalog []struct {
+			Type      string `json:"type"`
+			Endpoints []struct {
+				URL string `json:"url"`
+			} `json:"endpoints"`
+		} `json:"catalog"`
+	} `json:"token"`
+}
+
+func authenticatev3(authEndpoint, projectName, username, password string) (string, string, error) {
+	if !strings.HasSuffix(authEndpoint, "auth/tokens") {
+		if authEndpoint[len(authEndpoint)-1] == '/' {
+			authEndpoint = authEndpoint + "auth/tokens"
+		} else {
+			authEndpoint = authEndpoint + "/auth/tokens"
+		}
+	}
+	var authReq []byte
+	var err error
+	creds := &keystoneRequestV3{}
+	creds.Auth.Identity.Methods = []string{"password"}
+	creds.Auth.Identity.Password.User.Domain.Name = projectName
+	creds.Auth.Identity.Password.User.Name = username
+	creds.Auth.Identity.Password.User.Password = password
+	authReq, err = json.Marshal(creds)
+	if err != nil {
+		return "", "", fmt.Errorf("error encoding auth json request %s", err)
+	}
+	client := &http.Client{
+		Timeout: 30 * time.Minute,
+		Transport: &http.Transport{
+			MaxIdleConnsPerHost: 300,
+			MaxIdleConns:        0,
+			IdleConnTimeout:     5 * time.Second,
+			DisableCompression:  true,
+		},
+	}
+	resp, err := client.Post(authEndpoint, "application/json", bytes.NewBuffer(authReq))
+	if err != nil {
+		return "", "", fmt.Errorf("error making auth request %s", err)
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode/100 != 2 {
+		return "", "", fmt.Errorf("unexpected status code %d from auth", resp.StatusCode)
+	}
+	defer resp.Body.Close()
+	authToken := resp.Header.Get("X-Subject-Token")
+	if authToken == "" {
+		return "", "", fmt.Errorf("no X-Subject-Token in response")
+	}
+	var authResponse keystoneResponseV3
+	if err := json.NewDecoder(resp.Body).Decode(&authResponse); err != nil {
+		return "", "", fmt.Errorf("error decoding auth json response %s", err)
+	}
+	barbicanEndpoint := ""
+	for _, s := range authResponse.Token.Catalog {
+		if s.Type == "barbican" && len(s.Endpoints) > 0 {
```

Comment:
The endpoint type is barbican? Not something like keymanager?

Reply:
That's a good point. I'm not sure why I ended up with "barbican" there, but I see examples out there on the Internet using "keystore". https://github.com/cloudkeep/barbican/wiki/Barbican-Options:-authentication-with-Keystone

Reply:
I guess the only sad part about this is it probably has to be configurable...